### PR TITLE
shp::copy_async() local -> distributed + tests in both directions

### PR DESCRIPTION
**Pull request by mateuszpn**

### CODE REVIEW

**Comment by KFilipek** [comment]
Please remove the extra line in include/dr/shp/copy.hpp:147

File: include/dr/shp/copy.hpp [hunk 1]
```diff
@@ -105,6 +105,48 @@ OutputIt copy(InputIt first, InputIt last, OutputIt d_first) {
...
+    total_copied += n_to_copy;
+  }
+
+  sycl::queue q;
+
+  auto root_event = q.submit([=](auto &&h) { h.depends_on(events); });
+
+  return root_event;
+}
+
```
Comment (KFilipek) [inline include/dr/shp/copy.hpp:148]: Remove extra line

**Comment by lslusarczyk** [comment]
closes #14

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -41,9 +41,17 @@ repos:
   - id: doc8
 
 - repo: https://github.com/pycqa/isort
-  rev: 5.10.1
+  rev: 5.12.0
   hooks:
   - id: isort
+    args: ['--order-by-type', '--length-sort', "--profile", "black", "--filter-files"]
```
Comment (lslusarczyk) [inline .pre-commit-config.yaml:47]: I understand it was important to use higher version of isort in order to make code running on gdds.

Are below changes also needed? If they are not needed please remove.

File: include/dr/shp/copy.hpp [hunk 3]
```diff
@@ -105,6 +105,47 @@ OutputIt copy(InputIt first, InputIt last, OutputIt d_first) {
   return rv;
 }
 
+template <std::forward_iterator InputIt, lib::distributed_iterator OutputIt>
+cl::sycl::event copy_async(InputIt first, InputIt last, OutputIt d_first) {
```
Comment (lslusarczyk) [inline include/dr/shp/copy.hpp:109]: Once https://github.com/oneapi-src/distributed-ranges/pull/15 is merged please merge your branch to the latest main and add a test for this new _copy_async_ code.

**Review by lslusarczyk — CHANGES_REQUESTED** [review]
Unittest needed

File: .pre-commit-config.yaml (hunk 2 above)
Comment (mateuszpn) [reply .pre-commit-config.yaml:47]: Verified as not needed - removed

**Review by lslusarczyk — APPROVED** [review]

File: include/dr/shp/copy.hpp [hunk 4]
```diff
@@ -13,20 +13,22 @@
 
 namespace shp {
 
+template <typename Src, typename Dest>
+concept is_syclmemcopyable = std::is_same_v<std::remove_const_t<Src>, Dest> &&
```
Comment (BenBrock) [inline include/dr/shp/copy.hpp:17]: I recommend putting this inside an anonymous namespace.

```cpp
namespace {
}
```

Comment (BenBrock) [reply include/dr/shp/copy.hpp:17]: We should not be defining any symbols except for `inline` symbols, but I'm also fine with `__detail`.

Comment (rscohn2) [reply include/dr/shp/copy.hpp:17]: > We should not be defining any symbols except for `inline` symbols, but I'm also fine with `__detail`.

Duplication can also come from putting classes in anonymous namespaces. The implementations of the methods will not be shared across compilation units.

File: include/dr/shp/copy.hpp [hunk 5]
```diff
@@ -91,29 +74,67 @@ OutputIt copy(InputIt first, InputIt last, OutputIt d_first) {
...
+    auto &&segment = segments[segment_id];
+    std::size_t n_in_segment = segment.size();
+
+    std::size_t n_to_copy =
+        std::min<size_t>(n_in_segment, std::distance(first, last));
+
+    auto local_last = first;
+    std::advance(local_last, n_to_copy);
+
+    auto remote_iter = segment.begin();
```
Comment (lslusarczyk) [inline include/dr/shp/copy.hpp:118]: `remote_iter` variable is used in one place only. There is no point in making this variable. Just pass `segment.begin()` into `shp::copy_async` call (here and in the function above)

File: test/gtest/shp/copy.cpp [hunk 6]
```diff
@@ -4,42 +4,144 @@
...
+  using LocalVec = std::vector<typename AllocT::value_type>;
+};
 
-TEST(ShpTests, Copy_Dist2Local) {
-  const int n = 100;
-  std::size_t n_to_copy = 20;
+using AllocatorTypes =
+    ::testing::Types<shp::device_allocator<int>,
+                     shp::shared_allocator<long long unsigned int>>;
+TYPED_TEST_SUITE(CopyTest, AllocatorTypes);
```
Comment (rscohn2) [inline test/gtest/shp/copy.cpp:16]: This is nice. Can we do share tests for `shp::distributed_vector` and `mhp::distributed_vector` this way?

Comment (lslusarczyk) [reply test/gtest/shp/copy.cpp:16]: Yes, we can.
But let's do it in separate PR as it is not yet obvious to me how to do it in best way.

**Pull request merged**